sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/SynchronousMetricStorage.java

```diff
@@ -154,30 +164,173 @@ public void recordDouble(double value, Attributes attributes, Context context) {
   }
 
   @Override
-  public MetricData collectAndReset(long startEpochNanos, long epochNanos) {
+  public MetricDescriptor getMetricDescriptor() {
+    return metricDescriptor;
+  }
+
+  /**
+   * This method leverages the `savedDeltas` and `reportHistory` to construct a final metric point
+   * to report for this synchronous instrument.
+   */
+  @GuardedBy("collectLock")
+  @Nullable
+  private MetricData buildMetricFor(
+      CollectionHandle collector, long startEpochNanos, long epochNanos) {
+    Map<Attributes, T> result = new HashMap<>();
+    // Next merge the delta w/ the last set of points.
+    for (DeltaAccumulation<T> point : savedDeltas) {
+      if (!point.wasReadBy(collector)) {
+        mergeInPlace(result, point.read(collector));
+      }
+    }
+
+    long lastEpochNanos;
+    // First pull the last cumulative value.
+    if (reportHistory.containsKey(collector)) {
+      LastReportedAccumulation<T> last = reportHistory.get(collector);
+      // Send the accumulated deltas in w/ previous accumulation to get final result.
+      if (aggregator.isStateful()) {
+        mergeInPlace(result, last.getAccumlation());
+      }
+      lastEpochNanos = last.getEpochNanos();
+    } else {
+      lastEpochNanos = startEpochNanos;
+    }
+
+    // Now write the aggregated value back, and generate final metric.
+    reportHistory.put(collector, new LastReportedAccumulation<>(result, epochNanos));
+    // Don't make a metric if we have no values.
+    if (result.isEmpty()) {
+      return null;
+    }
+    return aggregator.toMetricData(result, startEpochNanos, lastEpochNanos, epochNanos);
+  }
+
+  /** Merges accumulations from {@code toMerge} into {@code result}. */
+  private void mergeInPlace(Map<Attributes, T> result, Map<Attributes, T> toMerge) {
+    toMerge.forEach(
+        (k, v) -> {
+          if (result.containsKey(k)) {
+            result.put(k, aggregator.merge(result.get(k), v));
+          } else {
+            result.put(k, v);
+          }
+        });
+  }
+
+  /** Removes deltas once all collectors have pulled them. */
+  @GuardedBy("collectLock")
+  private void cleanup(Set<CollectionHandle> collectors) {
+    Iterator<DeltaAccumulation<T>> i = savedDeltas.iterator();
+    while (i.hasNext()) {
+      DeltaAccumulation<T> delta = i.next();
+      if (delta.wasReadyByAll(collectors)) {
+        i.remove();
+      }
+    }
+    // TODO: Do we allow different pipelines?
+    Iterator<CollectionHandle> c = reportHistory.keySet().iterator();
+    while (c.hasNext()) {
+      if (!collectors.contains(c.next())) {
+        c.remove();
+      }
+    }
+  }
+
+  /** Collects bucketed metrics and resets the underlying storage for the next collection period. */
+  @Override
+  public MetricData collectAndReset(
```

Comment:
this is definitely a hairy method in a hairy class. We'll want to figure out how to refactor/cleanup to make it processable by normal humans at some point.

Reply:
(I see you've made some attempts at this already, so no criticism intended)

Reply:
totally agree!   If you help me with naming, I'll work on chunking this out into better pieces.   

To some extent, I think cleaning up the Aggregator interface can help a bit here.  To another extent, I went crazy with private data classes that could likely be better named.

Reply:
I think we can work on it over time. I don't think we necessarily need to block on this internal cleanup.

Reply:
I actually came up with something I kind of like. PTAL.

TL;DR:

- DeltaStorage keeps "live" synchronous aggregations and dumps delta-measurements on demand.
- TemporalStorage remembers "last" reported point for (used awkwardly for both delta + cumulative)
- MetricStorage classes now focus on pulling all components together, e.g. AttributesProcessor.